### PR TITLE
Migrate Edraak i18n platform

### PR DESCRIPTION
**Pull request by OmarIthawi**
Task: 
  - [Migrate edraak_i18n module with a feature flag](https://app.asana.com/0/256944342521670/276104526908811)

Things wasn't migrated:

 - Transifex theme command, will be migrated with the [Transifex task](https://app.asana.com/0/256944342521670/273208255426211).

### CODE REVIEW

File: cms/envs/aws.py [hunk 1]
```diff
@@ -452,6 +452,16 @@
...
 
+
+# Edraak Apps
+# Keep it in sync with {lms,cms}/envs/{test,aws}.py
+if FEATURES.get('EDRAAK_I18N_APP'):
+    # Common app, keep it in sync with the CMS
+    import edraak_i18n.helpers
+    INSTALLED_APPS += ('edraak_i18n',)
+    MIDDLEWARE_CLASSES = edraak_i18n.helpers.add_locale_middleware(MIDDLEWARE_CLASSES)
+
```
Comment (Salomari1987) [inline cms/envs/aws.py:426]: I like this

File: common/djangoapps/edraak_i18n/middleware.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    """
+    Changes the the language to `settings.LANGUAGE_CODE` for all non-API requests.
+
+    This will force the i18n machinery to always choose settings.LANGUAGE_CODE
+    as the default initial language, unless another one is set via sessions or cookies.
+
+    Should be installed *before* any middleware that checks request.META['HTTP_ACCEPT_LANGUAGE'],
+    specifically django.middleware.locale.LocaleMiddleware
+
+    Although the middleware is installed by default, it sill checks for the feature flag bellow in order to work.
```
Comment (Salomari1987) [inline common/djangoapps/edraak_i18n/middleware.py]: Spelling mistakes: 
* bellow is below (line 19)
* the the (line 11)

File: common/djangoapps/edraak_i18n/helpers.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        request: A django request.
+
+    Returns: True if the request is an API request and False otherwise.
+    """
+    if request.path.startswith('/api/'):
+        return True
+    elif request.path.startswith('/user_api/'):
+        return True
+    elif request.path.startswith('/notifier_api/'):
+        return True
```
Comment (Salomari1987) [inline common/djangoapps/edraak_i18n/helpers.py:52]: You can use regex here to reduce lines of code

File: cms/envs/test.py [hunk 4]
```diff
@@ -330,6 +330,17 @@
 # API access management -- needed for simple-history to run.
 INSTALLED_APPS += ('openedx.core.djangoapps.api_admin',)
 
+
+# Enable Edraak apps for testing
+# Keep it in sync with {lms,cms}/envs/{test,aws}.py
+import edraak_i18n.helpers
```
Comment (Salomari1987) [inline cms/envs/test.py]: use from ... import

File: common/djangoapps/edraak_i18n/tests.py [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+            'hello.world.middleware',
+        ))
+
+        self.assertIn('edraak_i18n.middleware.DefaultLocaleMiddleware', updated_middleware)
+
+        self.assertLess(
+            updated_middleware.index('edraak_i18n.middleware.DefaultLocaleMiddleware'),
+            updated_middleware.index('django.middleware.locale.LocaleMiddleware'),
+            'The i18n middleware should come before the `LocaleMiddleware`',
+        )
```
Comment (Salomari1987) [inline common/djangoapps/edraak_i18n/tests.py]: This is already tested before, no need to repeat again

File: common/djangoapps/edraak_i18n/helpers.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+    indexes = [
+        middleware_classes.index(class_name) for class_name in other_locale_middlewares
+        if class_name in middleware_classes
+    ]
+
+    first_index = min(indexes)
+
+    # Insert the DefaultLocaleMiddleware any other locale-related middleware in order for it to work
+    return middleware_classes[:first_index] + (edraak_middleware,) + middleware_classes[first_index:]
```
Comment (Salomari1987) [inline common/djangoapps/edraak_i18n/helpers.py]: You can do another logic with less complexity, check for the first element, if it is any of the (other_local_middlewares), then add the middleware as the first element, otherwise, make it the second, or you could always add it as the first, what do you think?

**Review by Salomari1987 — CHANGES_REQUESTED** [review]

File: cms/envs/test.py (hunk 4 above)
Comment (OmarIthawi) [reply cms/envs/test.py]: This is actually better, to avoid polluting the settings.

File: common/djangoapps/edraak_i18n/helpers.py (hunk 3 above)
Comment (OmarIthawi) [reply common/djangoapps/edraak_i18n/helpers.py:52]: Yes, but I'd like to be very specific and it is more readable for me this way.

File: cms/envs/aws.py (hunk 1 above)
Comment (OmarIthawi) [reply cms/envs/aws.py:426]: Glad you do 👍

File: common/djangoapps/edraak_i18n/helpers.py (hunk 6 above)
Comment (OmarIthawi) [reply common/djangoapps/edraak_i18n/helpers.py]: What about the `SessionMiddleware` or the `SafeSessionMiddleware`? locale middleware requires one of them to work.

Besides it's much [more defensive](http://scottdorman.github.io/2008/07/04/what-is-ldquodefensive-programmingrdquo/) to write it this way!

File: common/djangoapps/edraak_i18n/tests.py (hunk 5 above)
Comment (OmarIthawi) [reply common/djangoapps/edraak_i18n/tests.py]: 👍

**Comment by OmarIthawi** [comment]
All comments have been addressed. Please let me know if you have more.

File: common/djangoapps/edraak_i18n/helpers.py (hunk 6 above)
Comment (Salomari1987) [reply common/djangoapps/edraak_i18n/helpers.py]: But here you are not checking for ```SessionMiddleware``` or any other middleware that Locale middleware depends on, thus defensive programming does not exactly apply here, maybe if there was a function that checks middlewares and rearranges them correctly, what do you think?

File: common/djangoapps/edraak_i18n/helpers.py (hunk 3 above)
Comment (Salomari1987) [reply common/djangoapps/edraak_i18n/helpers.py:52]: its not generic though, if more urls are introduced in the future that represent api requests, then the decision tree becomes hideous, no?

**Comment by Salomari1987** [comment]
@OmarIthawi sounds good, maybe you can add a TODO: write more comprehensive tests

File: common/djangoapps/edraak_i18n/helpers.py (hunk 3 above)
Comment (OmarIthawi) [reply common/djangoapps/edraak_i18n/helpers.py:52]: It's not generic, and we'll wait for it to break in the future, we still rely on manual QA for e2e tests. 

While the regexp is a valid suggestion, it's not guaranteed to work.

For example edX could add any of these:

 - `/certificate_api/`
 - `/xmodule/api/`

There's still some guessing and the code won't be future proof anyway, in this case I always fallback to the stupid solution i.e. simple if statement vs. a RegExp.

**Comment by OmarIthawi** [comment]
> @OmarIthawi sounds good, maybe you can add a TODO: write more comprehensive tests

The edX default setup hates TODOs and the build will fail, so we won't write any TODOs.

If you have a suggestion for a specific missing test, please suggest. Otherwise, we'll go with what we have.

**Comment by OmarIthawi** [comment]
@Salomari1987 Let's wait for #39 to get merged so we can get rid off the failing flaky tests.

**Review by Salomari1987 — APPROVED** [review]

**Review by Salomari1987 — CHANGES_REQUESTED** [review]
Please update this branch with master

**Comment by OmarIthawi** [comment]
@Salomari1987 the tests should pass now.

### TL;DR

I've actually cherry-picked the flaky test fixups here as well. If either one is merged, we'll remove the cherry-picks from the others and rebase.

**Comment by Salomari1987** [comment]
@OmarIthawi PR Ratelimit is merged, so you can rebase here and reference me to merge

**Comment by OmarIthawi** [comment]
Thanks @Salomari1987. Just did.

**Comment by OmarIthawi** [comment]
@Salomari1987 it passed CI tests. 

!فرصة تراجعها قبل ما يغير رأيه

**Review by Salomari1987 — APPROVED** [review]

**Comment by OmarIthawi** [comment]
😄 الله يسعدك رفيق
